File: game/zone.go
// Package game 負責視野區域管理與視野內即時模擬。本檔為視野範圍內實體列表，對齊決策 004 視野內才即時模擬。
package game

// ViewRadius 為以角色為中心的視野半徑（格數），可調參數建議集中於 config。
const ViewRadius = 10

// ChunkSize 為單一區塊邊長（格數）；預設 151（22,801 格）。當前區塊常亮、區塊外黑、越界瞬間載入新區塊並關閉上個區塊。
const ChunkSize = 151

// InView 檢查 (ex,ey) 是否在 (cx,cy) 的視野範圍內（曼哈頓或歐幾里得；此處用簡單方形）。
func InView(cx, cy, ex, ey int) bool {
	dx := ex - cx
	if dx < 0 {
		dx = -dx
	}
	dy := ey - cy
	if dy < 0 {
		dy = -dy
	}
	return dx <= ViewRadius && dy <= ViewRadius
}

// ChunkIndex 回傳世界座標 (wx, wy) 所屬的區塊索引 (cx, cy)。區塊邊長為 ChunkSize；負座標用 floor 除法。
func ChunkIndex(wx, wy int) (cx, cy int) {
	if wx >= 0 {
		cx = wx / ChunkSize
	} else {
		cx = (wx - ChunkSize + 1) / ChunkSize
	}
	if wy >= 0 {
		cy = wy / ChunkSize
	} else {
		cy = (wy - ChunkSize + 1) / ChunkSize
	}
	return cx, cy
}

// ChunkBounds 回傳區塊 (cx, cy) 在世界座標上的範圍 [x0, x1)、[y0, y1)，含 x0 不含 x1，格數為 ChunkSize×ChunkSize。
func ChunkBounds(cx, cy int) (x0, y0, x1, y1 int) {
	x0 = cx * ChunkSize
	y0 = cy * ChunkSize
	x1 = x0 + ChunkSize
	y1 = y0 + ChunkSize
	return x0, y0, x1, y1
}

// InChunk 回傳世界格 (wx, wy) 是否落在區塊 (cx, cy) 內。
func InChunk(wx, wy, cx, cy int) bool {
	x0, y0, x1, y1 := ChunkBounds(cx, cy)
	return wx >= x0 && wx < x1 && wy >= y0 && wy < y1
}

File: src/game/zone.rs
// 視野區域與區塊索引幾何，對齊 Go game/zone.go。

/// 以角色為中心的視野半徑（格數）。
pub const VIEW_RADIUS: i32 = 10;

/// 單一區塊邊長（格數）；預設 151。
pub const CHUNK_SIZE: i32 = 151;

/// 檢查 `(ex,ey)` 是否在 `(cx,cy)` 的方形視野內。
#[must_use]
pub fn in_view(cx: i32, cy: i32, ex: i32, ey: i32) -> bool {
    let dx = (ex - cx).abs();
    let dy = (ey - cy).abs();
    dx <= VIEW_RADIUS && dy <= VIEW_RADIUS
}

/// 世界座標 `(wx, wy)` 所屬區塊索引 `(cx, cy)`；負座標用 floor 除法。
#[must_use]
pub fn chunk_index(wx: i32, wy: i32) -> (i32, i32) {
    let cx = if wx >= 0 {
        wx / CHUNK_SIZE
    } else {
        (wx - CHUNK_SIZE + 1) / CHUNK_SIZE
    };
    let cy = if wy >= 0 {
        wy / CHUNK_SIZE
    } else {
        (wy - CHUNK_SIZE + 1) / CHUNK_SIZE
    };
    (cx, cy)
}

/// 區塊 `(cx,cy)` 在世界座標上的範圍 `[x0,x1)×[y0,y1)`。
#[must_use]
pub fn chunk_bounds(cx: i32, cy: i32) -> (i32, i32, i32, i32) {
    let x0 = cx * CHUNK_SIZE;
    let y0 = cy * CHUNK_SIZE;
    let x1 = x0 + CHUNK_SIZE;
    let y1 = y0 + CHUNK_SIZE;
    (x0, y0, x1, y1)
}

/// 世界格 `(wx, wy)` 是否落在區塊 `(cx, cy)` 內。
#[must_use]
pub fn in_chunk(wx: i32, wy: i32, cx: i32, cy: i32) -> bool {
    let (x0, y0, x1, y1) = chunk_bounds(cx, cy);
    wx >= x0 && wx < x1 && wy >= y0 && wy < y1
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn in_view_center() {
        assert!(in_view(0, 0, 5, -3));
        assert!(!in_view(0, 0, 11, 0));
    }

    #[test]
    fn chunk_index_negative() {
        assert_eq!(chunk_index(-1, 0), (-1, 0));
        assert_eq!(chunk_index(-151, 0), (-1, 0));
        assert_eq!(chunk_index(-152, 0), (-2, 0));
    }

    #[test]
    fn in_chunk_bounds() {
        let (cx, cy) = (1, 2);
        let (x0, y0, x1, y1) = chunk_bounds(cx, cy);
        assert!(in_chunk(x0, y0, cx, cy));
        assert!(!in_chunk(x1, y0, cx, cy));
        assert!(!in_chunk(x0, y1, cx, cy));
    }
}
